Plugin configuration

bobwhitelock opened the pull request:
This PR adds support for configuring enabled plugins via the standard `metal configure` process. It's based on https://github.com/alces-software/metalware/pull/291 so take a look at that first.

The key commit here is https://github.com/alces-software/metalware/commit/6512013b8541a87f1b84452b6a09b5e6a0facf2d, which contains the main change to loading of configure questions and a rough test for this, as well as an overall description of how this works. Later commits then mostly contain refactoring to tidy things up following this change, apart from https://github.com/alces-software/metalware/commit/c7644f15e0cc432664b28217021b464918e40cc6 which adds handling for the situation where no `configure.yaml` file exists for a plugin. So it would be particularly useful if you could take a look at these two commits and tell me if the main changes here seem reasonable; feel free to look at the other refactoring and let me know what you think as well if you like.

bobwhitelock commented:
Rebased on `develop` for you to review and to attempt to fix that intermittent CI test failure.

WilliamMcCumstie commented on spec/filesystem.rb:
```diff
@@ -38,9 +38,10 @@ class FileSystem
+  delegate :mkdir_p, :touch, :rm_rf, to: FileUtils
   delegate :write, to: File
   delegate :dump, to: Metalware::Data
+  delegate :enable!, to: Metalware::Plugins
```
I don't think `enable!` should be delegated to `Metalware::Plugins`. I would expect `FileSystem.enable!` to enable `FakeFS` as a whole and not enable the plugins. A more descriptive method would be `enable_plugin`.

Also I don't think this needs to be a `!` method unless an additional "safe" method is also specified. There are plenty of cases where ruby methods are "dangerous" without having the `!` (e.g. `array.delete`).

bobwhitelock replied:
Addressed your first point in https://github.com/alces-software/metalware/commit/a40c5d8448df93d53785929288e5f9af9abe10ad; and extracted these setup methods to their own module in https://github.com/alces-software/metalware/commit/db092b03bf7eaf99e61c89df8473e2746e83b612.

Regarding the second point, I'm not sure - I try to define `!` methods even if there isn't a corresponding 'safe' method, for any situation which will potentially unexpectedly: (a) effect things on disk, (b) mutate data, or (c) normally raise an exception. In this case this does both (a) and (b) so I find it helpful to know that I may need to be a bit careful when calling this method.

In general I find this works reasonably for my code as I try to minimize the places which will do these things and so the `!` methods stand out a bit, but it can be a bit fuzzy when using a `!` in a method will be helpful. I'm open to changing my opinion on this however :slightly_smiling_face:.

If we do decide to change this, this is an issue which applies more generally across Metalware, as we have various places where we define `!` methods without any corresponding `!`-less methods; for instance:

```
$ git grep -E 'def [a-z_]+!'
spec/filesystem.rb:  def debug!
src/cli.rb:    def run!
src/commander_extensions.rb:    def validate_syntax!
src/commander_extensions.rb:    def validate_correct_number_of_args!(args)
src/commands/hunter.rb:      def listen!
src/gui/bin/setup:def system!(*args)
src/gui/bin/update:def system!(*args)
src/hash_mergers/hash_merger.rb:      def replace_file_with_same_basename!(files_namespace, file_identifier)
src/plugins.rb:      def enable!(plugin_name)
src/plugins.rb:      def disable!(plugin_name)
src/plugins.rb:      def validate_plugin_exists!(plugin_name)
src/plugins.rb:      def update_enabled_plugins!(new_enabled_plugins)
src/plugins.rb:    def enable!
```

WilliamMcCumstie replied:
Okay checking the ruby doc the convention is the `!` method is a dangerous method. However it also specifies that there is almost always a safe method for every `!`. I'm happy with it as is, metalware has been playing hard and fast with this convention anyway.

bobwhitelock replied:
OK great, thanks for pointing it out though; it's something to consider in future.

WilliamMcCumstie commented on spec/validation/loader_spec.rb:
```diff
@@ -47,9 +47,66 @@
+    let :example_plugin_configure_questions_hash do
+      configure_sections.map do |section|
+        [
+          section, [{
```
Can this be broken up to something like:
```
key = section
value = **blah**
[key, value]
```
In it's current form it isn't obvious how this will be converted into an array due to the size complexity of the array object.

WilliamMcCumstie commented on spec/validation/loader_spec.rb:
```diff
@@ -47,9 +47,66 @@
+      end.to_h
+    end
+
     let :filesystem do
```
Replace with `begin do FileSystem.root_setup do .... end end` notation. This removes the need to go `FileSystem.test` for each test.

bobwhitelock replied:
Yep, done in https://github.com/alces-software/metalware/commit/691ac8706964c7b5fd0220e30d8e82bcc097d012.

WilliamMcCumstie commented on src/validation/loader.rb:
```diff
@@ -71,12 +71,33 @@ def answer(absolute_path, section)
       end
 
+      # XXX Extract object for loading configure data?
       def configure_data_tree
```
This can probably be renamed to `configure_tree`. The `data` is superfluous at this point.

bobwhitelock replied:
Good point - though actually looking at this again the existence of this method seems a bit pointless, as it's just called from `configure_data` and is the only thing that does. Therefore I've inlined the whole thing in https://github.com/alces-software/metalware/commit/2df9b97764c32129088b64a3c348a6afb1a5ce1a.

bobwhitelock commented on spec/validation/loader_spec.rb:
```diff
@@ -49,19 +49,17 @@
 
     let :filesystem do
```
I was initially a bit unsure about using this method as I thought it might be a bit too magic, but actually this is quite nice and avoids having to remember to wrap each test case in a `filesystem.test` block. Thanks for the suggestion, made in https://github.com/alces-software/metalware/commit/691ac8706964c7b5fd0220e30d8e82bcc097d012.

WilliamMcCumstie replied:
The use of the `root_setup` is a bit clunky. Every now and then I forget to put it in a `:before` block which can cause some weird errors depending on which order the tests run in. This could probably be improved in the future.

bobwhitelock replied:
Yep, there's a few things we could still improve about the `FileSystem`/`FakeFS` etc. stuff - it's much better than it used to be however and using `root_setup` here is an improvement IMO :slightly_smiling_face:.

bobwhitelock commented on src/plugins.rb:
```diff
@@ -83,6 +83,7 @@ def all_plugin_names
       end
 
       def plugin_directories
+        return [] unless plugins_dir.exist?
```
It can. Not sure if this would be an improvement though? We basically have the normal code path and the special short circuit case for when a repo lacks the `plugins/` dir. Not sure if an alternative way of writing the same thing would be an improvement, and I don't think it matters much either way.

WilliamMcCumstie replied:
Doesn't super matter, it just gives me the warm and fuzzies when ever I can do some complex ruby code in a single line.

bobwhitelock replied:
lol - have you considered taking up Perl? :grimacing:

bobwhitelock commented on src/plugins.rb:
```diff
@@ -118,8 +118,19 @@ def enable!
+      new(plugin).build
+    end
+
+    def build
```
I think having the internal build method is useful:

a. it's descriptive of what we're doing, rather than being a generic name;

b. in general it's expected that `SomeClass .new` will give an instance of `SomeClass`, not some other object, but in this case we want to return a hash, and not an instance of `ConfigureQuestionsBuilder`; the instance of this only exists while `ConfigureQuestionsBuilder.build` is executing to encapsulate the methods and state involved in this process. Therefore I think this change would be surprising.

WilliamMcCumstie replied:
Ok good point, it can remain as is then.

bobwhitelock commented on src/plugins.rb:
```diff
@@ -116,5 +117,56 @@ def enabled_identifier
+
+    private
+
+    def question_tree_for_section(section)
```
Good point; done in https://github.com/alces-software/metalware/commit/c769623986d1e7def9117c34a35ed437ff1d9724.

bobwhitelock commented on src/validation/loader.rb:
```diff
@@ -71,12 +71,33 @@ def answer(absolute_path, section)
+        [
+          repo_configure_questions,
+          *plugin_configure_questions,
+        ].flat_map do |question_group|
```
Well why do we need anything? :slightly_smiling_face: Ruby provides us with several ways to do most things and it's up to us to choose the 'best'/most appropriate for the situation. In this case `flat_map` is a standard method to `map` over an `Enumerable` and then `flatten` the result with one method call. I'd argue that my way is slightly better as:

1. it only involves 1 method call rather than 2;

2. it doesn't involve any mutation, which all things being equal is slightly superior to mutating things in place (as that can have surprising effects).

But there's not much in it either way, they're both pretty simple :slightly_smiling_face:.

WilliamMcCumstie replied:
Okay, I guess the thing that threw me is the `flat_map` implies that we have an object that needs to be flattened. Does the method also iterate through the `dependent` questions correctly? If so then it seems appropriate.

However if it is only the result of how the arrays are combined, then it is sightly confusing.

bobwhitelock replied:
So briefly what this is doing is:

1. We get an array containing the question hashes for the repo, and each enabled plugin in turn: `[repo_configure_questions, *plugin_configure_questions]` (this is identical to your use of `concat` just without mutating the array);

2. We map over this to get an array containing arrays for every question in the repo and every enabled plugin for the given configure section;

3. the `flatten` occurs as part of the `flat_map` to turn this array of arrays into 1 array containing e.g. every domain question from the repo and every domain question from every enabled plugin in order.

So a `flatten` will always be needed in some way, or at least some equivalent of this will be, in order to turn these nested arrays into one big array of question hashes; the same would be needed in your `concat` version as well I believe.

Hope this helps :slightly_smiling_face:.

bobwhitelock commented:
@WilliamMcCumstie addressed your comments/justified why I've done things my way. Would be great if you could take another look and let me know what you think.

WilliamMcCumstie commented:
This looks pretty good. If you could look at https://github.com/alces-software/metalware/pull/292#discussion_r161581453 that would be great. Otherwise it should be ready to merge.

bobwhitelock commented:
Great, thanks for the review; merging now :shipit:
